### PR TITLE
Add tests for r_id_storage

### PR DESCRIPTION
**Pull request by condret**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: test/unit/test_idstorage.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	ut32 id;
+	r_id_storage_add (ids, str, &id);
+	void *ptr = r_id_storage_get (ids, id);
+	r_id_storage_free (ids);
+	mu_assert_ptreq (ptr, str, "id_storage_add");
+	mu_end;
+}
+
+bool test_r_id_storage_set(void) {
+	char *str = "lol";
```
Comment (trufae) [inline test/unit/test_idstorage.c:16]: This is const

File: test/unit/test_idstorage.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	r_id_storage_add (ids, "lol", &id);
+	r_id_storage_delete (ids, id);
+	void *ptr = r_id_storage_get (ids, id);
+	r_id_storage_free (ids);
+	mu_assert_ptreq (ptr, NULL, "id_storage_delete");
+	mu_end;
+}
+
+bool test_r_id_storage_take0(void) {
+	char *str = "lol";
```
Comment (trufae) [inline test/unit/test_idstorage.c:37]: Same

File: test/unit/test_idstorage.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	mu_run_test (test_r_id_storage_add);
+	mu_run_test (test_r_id_storage_set);
+	mu_run_test (test_r_id_storage_delete);
+	mu_run_test (test_r_id_storage_take0);
+	mu_run_test (test_r_id_storage_take1);
+	return tests_passed != tests_run;
+}
+
+int main(int argc, char **argv) {
+	return all_tests();
```
Comment (trufae) [inline test/unit/test_idstorage.c:70]: Missing space

File: test/unit/test_idstorage.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#include <r_util.h>
+#include "minunit.h"
+
+bool test_r_id_storage_add(void) {
+	char *str = "lol";
+	RIDStorage *ids = r_id_storage_new (5, 23);
+	ut32 id;
+	r_id_storage_add (ids, str, &id);
```
Comment (trufae) [inline test/unit/test_idstorage.c:8]: Check return value

File: test/unit/test_idstorage.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	r_id_storage_free (ids);
+	mu_assert_ptreq (ptr, str, "id_storage_set");
+	mu_end;
+}
+
+bool test_r_id_storage_delete(void) {
+	RIDStorage *ids = r_id_storage_new (5, 23);
+	ut32 id;
+	r_id_storage_add (ids, "lol", &id);
+	r_id_storage_delete (ids, id);
```
Comment (trufae) [inline test/unit/test_idstorage.c:29]: Check return value deleting a valid and an invalid fd

File: test/unit/test_idstorage.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	mu_assert_ptreq (ptr, NULL, "id_storage_delete");
+	mu_end;
+}
+
+bool test_r_id_storage_take0(void) {
+	char *str = "lol";
+	RIDStorage *ids = r_id_storage_new (5, 23);
+	ut32 id;
+	r_id_storage_add (ids, str, &id);
+	void *ptr = r_id_storage_take (ids, id);
```
Comment (trufae) [inline test/unit/test_idstorage.c:41]: Difference between get and take?

File: test/unit/test_idstorage.c [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	r_id_storage_free (ids);
+	mu_assert_ptreq (ptr, str, "id_storage_take 0");
+	mu_end;
+}
+
+bool test_r_id_storage_take1(void) {
+	char *str = "lol";
```
Comment (trufae) [inline test/unit/test_idstorage.c:48]: Const

File: test/unit/test_idstorage.c (hunk 4 above)
Comment (trufae) [reply test/unit/test_idstorage.c:8]: It is valid to add NULL?

Comment (trufae) [reply test/unit/test_idstorage.c:8]: Do we have an R_IO_INVALID_ID UT32_MAX?

**Review by trufae — COMMENTED** [review]
Check the comments. Looks good so far

File: test/unit/test_idstorage.c (hunk 5 above)
Comment (condret) [reply test/unit/test_idstorage.c:29]: this would be 2 more tests, ugh

Comment (condret) [reply test/unit/test_idstorage.c:29]: delete is void. why would you care if delete returns success indicator? This test should test if an entry exists after deletion.

File: test/unit/test_idstorage.c (hunk 4 above)
Comment (condret) [reply test/unit/test_idstorage.c:8]: Adding NULL to the storage can be done in theory, but you should do that if you want to use iteration functions

Comment (condret) [reply test/unit/test_idstorage.c:8]: and yes that limitiation of max io id exists, but it's not part of this test

File: test/unit/test_idstorage.c (hunk 6 above)
Comment (condret) [reply test/unit/test_idstorage.c:41]: get gets you the element that is adressed by the id.
take does the same but also deletes the element from the idstorage

File: test/unit/test_idstorage.c (hunk 5 above)
Comment (trufae) [reply test/unit/test_idstorage.c:43]: To ensure the element u tried to delete was there already. Checking and deleting can be racy

File: test/unit/test_idstorage.c (hunk 6 above)
Comment (trufae) [reply test/unit/test_idstorage.c:55]: Ok thats what i was expecting too but is the only place in r2land where this verb is used

File: test/unit/test_idstorage.c [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+bool test_r_id_storage_delete(void) {
+	RIDStorage *ids = r_id_storage_new (5, 23);
+	ut32 id;
+	r_id_storage_add (ids, "lol", &id);
+	r_id_storage_delete (ids, id);
+	void *ptr = r_id_storage_get (ids, id);
+	r_id_storage_free (ids);
+	mu_assert_ptreq (ptr, NULL, "id_storage_delete");
```
Comment (ret2libc) [inline test/unit/test_idstorage.c:46]: you could use `mu_assert_null`

**Review by ret2libc — APPROVED** [review]
Thanks!

File: test/unit/test_idstorage.c (hunk 5 above)
Comment (condret) [reply test/unit/test_idstorage.c:43]: this whole api is racy, no mutexes anywhere. we can make this threadsafe at some point, but that would be out of scope of this pr.
What I want to say is, I'm fine with changing delete. but I don't want to do this now. Feel free to modify the pr, if you really think this should be part of this pr

**Review by XVilka — APPROVED** [review]

**Pull request merged**